balancer-js/src/modules/pools/model-provider.ts

```diff
@@ -0,0 +1,124 @@
+import type { BalancerDataRepositories, Pool, PoolModel } from '@/types';
+import { PoolApr } from './apr/apr';
+import { Liquidity } from '../liquidity/liquidity.module';
+
+// Can be changed outside for dependent services rate-limits
+// eslint-disable-next-line prefer-const
+export let POOLS_PER_PAGE = 10;
+
+/**
+ * Use-cases layer for generating live pools data
+ */
+export class ModelProvider {
+  constructor(private repositories: BalancerDataRepositories) {}
+
+  static async resolve(model: PoolModel): Promise<Pool> {
+    return {
+      ...model,
+      apr: await (async () => {
+        try {
+          const apr = await model.calcApr();
+          return apr;
+        } catch (e) {
+          console.log(e);
+          return;
+        }
+      })(),
+    };
+  }
+
+  static wrap(data: Pool, repositories: BalancerDataRepositories): PoolModel {
+    return {
+      ...data,
+      calcLiquidity: async function () {
+        const liquidityService = new Liquidity(
+          repositories.pools,
+          repositories.tokenPrices
+        );
+
+        return liquidityService.getLiquidity(this);
+      },
+      calcApr: async function () {
+        const aprService = new PoolApr(
+          data,
```

Comment:
Should be `this` instead of data to ensure this always uses the latest data.

Reply:
That looks good. I don't think the pool instance should have functions to calculate APR / Liquidity etc if it's going to use stale data, as that will cause a lot of confusion. Plus it doesn't really make sense for a class to have a function that doesn't use the class data for it's calculation.